File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        gifts = [-gift for gift in gifts]  # Max-heap using negative values
        heapq.heapify(gifts)
        
        while k > 0:
            max_gifts = -heapq.heappop(gifts)
            if max_gifts == 1:
                heapq.heappush(gifts, -1)  # Correctly add 1 back to the heap
                break
            leave = math.isqrt(max_gifts)
            heapq.heappush(gifts, -leave)
            k -= 1
        
        return -sum(gifts)

File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int gift : gifts) {
            pq.add(gift);
        }
        
        while (!pq.isEmpty() && k-- > 0) {
            int maxGifts = pq.poll();
            if (maxGifts == 1)
            {
                pq.add(1);
                break;                
            } 
            int leave = (int) Math.sqrt(maxGifts);
            pq.add(leave);
        }
        
        long ans = 0;
        while (!pq.isEmpty()) {
            ans += pq.poll();
        }
        
        return ans;
    }
}
